compiler/src/iree/compiler/Codegen/Common/GPU/GPUMaterializeEncoding.cpp

```diff
@@ -340,6 +340,95 @@ struct GPUSetEncodingOpLoweringConversion
+          getTypeConverter()->convertType(unsetEncodingOp.getResultType());
+      if (targetType != result.getType()) {
+        result = rewriter.create<tensor::CastOp>(unsetEncodingOp.getLoc(),
+                                                 targetType, result);
+      }
+      rewriter.replaceOp(unsetEncodingOp, result);
+      return success();
+    }
+
+    Location loc = unsetEncodingOp.getLoc();
+
+    FailureOr<MaterializeEncodingInfo> maybeEncodingInfo =
+        converter->getEncodingInfo(unsetEncodingOp.getSource().getType());
+    if (failed(maybeEncodingInfo)) {
+      return rewriter.notifyMatchFailure(unsetEncodingOp,
+                                         "unhandled result encoding");
+    }
+    SmallVector<int64_t> innerTiles = maybeEncodingInfo->innerTileSizes;
+    SmallVector<int64_t> intrinsicVectorShape =
+        maybeEncodingInfo->innerTileShapes;
+
+    rewriter.setInsertionPoint(*unPackOp);
+
+    auto sourceShape = unPackOp->getSourceType().getShape();
+    auto iT1 = intrinsicVectorShape[0];
+    auto iT2 = intrinsicVectorShape[1];
+    auto oT1 = sourceShape[2] / iT1;
+    auto oT2 = sourceShape[3] / iT2;
+    SmallVector<int64_t> transposeSourceDims = {
+        sourceShape[0], sourceShape[1], oT1, iT1, oT2, iT2};
+    assert(transposeSourceDims.size() == 6);
+
+    size_t targetRank = unsetEncodingOp.getResultType().getRank();
+
+    SmallVector<int64_t> transposePerm;
+    transposePerm.push_back(0);
+    transposePerm.push_back(1);
+    for (auto perm : maybeEncodingInfo->permutation) {
+      transposePerm.push_back(targetRank + perm);
+    }
+    SmallVector<int64_t> expandShapeResultDims = transposeSourceDims;
+    applyPermutationToVector(expandShapeResultDims, transposePerm);
+
+    auto expandShapeResultType = RankedTensorType::get(
+        expandShapeResultDims, unPackOp->getSourceType().getElementType());
+
+    tensor::ExpandShapeOp expandShapeOp =
+        reshapeForInnerDims(rewriter, unPackOp->getSource(),
+                            expandShapeResultType, unPackOp->getDestRank());
+
+    auto emptyTensor = rewriter.create<tensor::EmptyOp>(
+        loc, transposeSourceDims,
+        unsetEncodingOp.getSourceType().getElementType());
+    auto transposeOp = rewriter.create<linalg::TransposeOp>(
+        loc, expandShapeOp, emptyTensor,
+        invertPermutationVector(transposePerm));
+
+    auto transposeResultType = RankedTensorType::get(
+        transposeSourceDims, unsetEncodingOp.getSourceType().getElementType());
+    std::optional<SmallVector<ReassociationIndices>> collapseReassoc =
+        getReassociationIndicesForReshape(transposeResultType,
+                                          unPackOp->getSourceType());
+    assert(collapseReassoc.has_value());
+    auto collapseShapeOp = rewriter.create<tensor::CollapseShapeOp>(
+        loc, unPackOp->getSourceType(), transposeOp->getResult(0),
+        *collapseReassoc);
+    unPackOp->setOperand(0, collapseShapeOp.getResult());
```

Comment:
Can we just replace the collapse shape op with unsetEncodingOp rather than setting the operand? I mean, the SSA def chain would be obeyed anyway.

Reply:
I think we still want to keep the unpack op?

Reply:
I think the issue is that there are `expand_shape + transpose` right before the unpack, but @lialan creates the unpack op first. We don't need this change if you move the the unpack op creation after the transpose op and pass the result of transposition to the method directly.

E.g.,

```cpp 
    auto unPackOp = lowerUnsetEncodingToUnpackOp(
        rewriter, unsetEncodingOp, transpose->getResult(0), *converter,
        this->materializeEncodingValueFn);
    if (failed(unPackOp)) {
      Value result = adaptor.getSource();
      Type targetType =
          getTypeConverter()->convertType(unsetEncodingOp.getResultType());
      if (targetType != result.getType()) {
        result = rewriter.create<tensor::CastOp>(unsetEncodingOp.getLoc(),
                                                 targetType, result);
      }
      rewriter.replaceOp(unsetEncodingOp, result);
      return success();
    }
```

Note that you need to pass `adaptor.getSource()` to the first op of the chain. The accessor from `adaptor` returns the converted operands to you. The accessor from the `unset_encoding` op returns the non-converted operands to you. This is how dialect conversion works.

Reply:
done.